### PR TITLE
Fix theme import

### PR DESCRIPTION
**Pull request by hecon5**
Fix theme import and add logging to import/export routines to aid finding errors or issues.

### CODE REVIEW

**Comment by hecon5** [comment]
@joyfullservice I think this is ready; it takes a slightly different tack for error handling, but hopefully will allow to track down the actual progenitor of the error.

Let me know your thoughts, or if you have an alternate idea on how better to handle this.

**Comment by joyfullservice** [comment]
I like the general approach here. I will leave some more specific comments inline with the code...

File: Version Control.accda.src/modules/modImportExport.bas [hunk 1]
```diff
@@ -17,6 +18,9 @@ Option Explicit
 '---------------------------------------------------------------------------------------
 '
 Public Sub ExportSource(blnFullExport As Boolean)
+    On Error Resume Next
+    Dim FunctionName as String
+    FunctionName = moduleName & "ExportSource:blnFullExport:" & CStr$(blnFullExport)
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:23]: Same as noted above. It is a neat idea to log the parameters in play, but relevant items should be added to the _description_ rather than the _source_. Also note that the log file already records whether the export is being performed with the FastSave option, which might be what you are wanting to know from this parameter.

File: Version Control.accda.src/modules/modImportExport.bas [hunk 2]
```diff
@@ -107,6 +111,8 @@ Public Sub ExportSource(blnFullExport As Boolean)
                 Log.Increment
                 Log.Add "  " & cDbObject.Name, Options.ShowDebug
                 cDbObject.Export
+                CatchAny eelError, Err.Number & ":" & Err.Description, FunctionName & ":Exporting:" & _
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:114]: Good spot to add the `CatchAny` call. Note that there should be a space after the colon to separate the description from the error number. 😄 (And other adjustments described above with regard to the source.)

File: Version Control.accda.src/modules/modImportExport.bas [hunk 3]
```diff
@@ -144,8 +150,12 @@ Public Sub ExportSource(blnFullExport As Boolean)
     Perf.EndTiming
     Log.Add vbCrLf & Perf.GetReports, False
     
+    If Log.ErrorLevel = eelNoError Then 
+        Log.SaveFile FSO.BuildPath(Options.GetExportFolder, "Export.log")
+    Else
     ' Save log file to disk
-    Log.SaveFile FSO.BuildPath(Options.GetExportFolder, "Export.log")
+        Log.SaveFile FSO.BuildPath(Options.GetExportFolder, "Export-Errors-" & Format(Now, "YYYY-MM-DD-hh.mm.ss") & ".log")
+    End If
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:158]: I am not sure that I fully understand why we would want the additional log files to be created here... It seems like a typical usage scenario would involve exporting source, reviewing the log for errors, resolving problems, exporting again, etc... Additional log files would require git ignore patterns, etc... to keep from cluttering up commits.

If you need to preserve an error log from a specific export, wouldn't you just rename and save the file? For most users in day-to-day operations, I am thinking that simple is better with a single log file, and let them manually save other copies if they want.

File: Version Control.accda.src/modules/modImportExport.bas [hunk 4]
```diff
@@ -175,6 +185,9 @@ End Sub
 '---------------------------------------------------------------------------------------
 '
 Public Sub Build(strSourceFolder As String, blnFullBuild As Boolean)
+    On Error Resume Next
+    Dim FunctionName as String
+    FunctionName = moduleName & "Build:blnFullBuild:" & CStr$(blnFullBuild)
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:190]: Same as above...

File: Version Control.accda.src/modules/modImportExport.bas [hunk 5]
```diff
@@ -336,8 +349,12 @@ Public Sub Build(strSourceFolder As String, blnFullBuild As Boolean)
...
                 If blnFullBuild Then
                     cCategory.Import CStr(varFile)
+                    CatchAny eelError, Err.Number & ":" & Err.Description, FunctionName & ":Importing:" & _
+                        Options.UseFastSave & LCase(cCategory.Category) & " " & FSO.GetFileName(varFile), True, True
                 Else
                     cCategory.Merge CStr(varFile)
+                    CatchAny eelError, Err.Number & ":" & Err.Description, FunctionName & ":Merging:" & _
+                        Options.UseFastSave & LCase(cCategory.Category) & " " & FSO.GetFileName(varFile), True, True
                 End If
             Next varFile
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:359]: I think you could just use a single logger for this after the `End If`, with an embedded `IIF()` function to add the type to the description. I.e. `CatchAny eelError, Err.Number & ":" & Err.Description & " While " & IIF(blnFullBuild, "building ","merging ") & varfile`...

File: Version Control.accda.src/modules/modImportExport.bas [hunk 6]
```diff
@@ -374,8 +391,12 @@ Public Sub Build(strSourceFolder As String, blnFullBuild As Boolean)
     Perf.EndTiming
     Log.Add vbCrLf & Perf.GetReports, False
     
+    If Log.ErrorLevel = eelNoError Then 
+        Log.SaveFile FSO.BuildPath(Options.GetExportFolder, IIf(blnFullBuild, "Import", "Merge") & ".log")
+    Else
     ' Write log file to disk
-    Log.SaveFile FSO.BuildPath(Options.GetExportFolder, IIf(blnFullBuild, "Import.log", "Merge.log"))
+        Log.SaveFile FSO.BuildPath(Options.GetExportFolder, IIf(blnFullBuild, "Import-Error", "Merge-Error") & Format(Now, "-YYYY-MM-DD-hh.mm.ss") & ".log")
+    End If
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:399]: See discussion above regarding the additional log files.

**Review by joyfullservice — COMMENTED** [review]
I added a number of comments to the proposed changes. This may seem like a lot of issues to highlight, but keep in mind that we are also establishing a precedent on how we flesh out the error handling as we go forward. 😄 You are doing a great job! After these issues are worked through, I think we are pretty close to getting this pushed through to the `dev` branch! 👍

File: Version Control.accda.src/modules/clsDbTheme.bas [hunk 7]
```diff
@@ -51,13 +51,23 @@ Private Sub IDbComponent_Export()
...
+
     ' Query theme file details
     strSql = "SELECT [Data] FROM MSysResources WHERE [Name]='" & m_Name & "' AND Extension='" & m_Extension & "'"
     Set m_Dbs = CurrentDb
     Set rst = m_Dbs.OpenRecordset(strSql, dbOpenSnapshot, dbOpenForwardOnly)
     
     ' If we get multiple records back we don't know which to use
-    If rst.RecordCount > 1 Then Err.Raise 42, , "Multiple records in MSysResources table were found that matched name '" & m_Name & "' and extension '" & m_Extension & "' - Compact and repair database and try again."
+    If rst.RecordCount > 1 Then 
+        Err.Raise 42, , "Multiple records in MSysResources table were found that matched name '" & _
```
[the remark this answers is not in the document]
Comment (hecon5) [reply Version Control.accda.src/modules/clsDbTheme.bas:65]: That's what I thought, but since err.raise was already there #NotMe, I just left it in case there was a reason we were causing err 42. Didn't want to forget my towel...

File: Version Control.accda.src/modules/clsDbTheme.bas [hunk 8]
```diff
@@ -51,13 +51,23 @@ Private Sub IDbComponent_Export()
     Dim rstAtc As Recordset2
     Dim strSql As String
     
+    On Error Resume Next
+    Dim FunctionName As String
+    FunctionName = moduleName & "IDbComponent_Export:"
```
[the remark this answers is not in the document]
Comment (hecon5) [reply Version Control.accda.src/modules/clsDbTheme.bas:56]: To make sure I got it: use it like this, instead:

```VBA 
 CatchAny eelCritical, Err.Number & ":" & Err.Description, moduleName & ".Export:InputInfo:" & data-to-be-logged


CatchAny eelCritical, Err.Number & ":" & Err.Description, moduleName & ".Import:InputInfo:" & data-to-be-logged
```

File: Version Control.accda.src/modules/modImportExport.bas (hunk 3 above)
Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:158]: This may be a bad programmer habit, but I tend to just export a bunch, and move on. I've also had a few cases of intermittent errors that didn't happen every time, and since Export/import logs are overwritten when they go fine, making a separate one that wasn't made it easier to figure out what error happened on export click 22.

Since `Export.log` and `import.log` are already ignored (they should be as I don't need to see another identical log and have this clutter stuff up!), this has caused me to notice "oh, hey, that thing I walked away from for more coffee/a stretch didn't work as I intended, I should look at it."

File: Version Control.accda.src/modules/modImportExport.bas (hunk 5 above)
Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:359]: ooh, that could work; was trying to keep it light and not need to process any iff/else, even though that would expand the rewrite code base if CatchAll were to be adjusted.

File: Version Control.accda.src/modules/clsDbTheme.bas [hunk 9]
```diff
@@ -51,13 +51,23 @@ Private Sub IDbComponent_Export()
...
     strSql = "SELECT [Data] FROM MSysResources WHERE [Name]='" & m_Name & "' AND Extension='" & m_Extension & "'"
     Set m_Dbs = CurrentDb
     Set rst = m_Dbs.OpenRecordset(strSql, dbOpenSnapshot, dbOpenForwardOnly)
     
     ' If we get multiple records back we don't know which to use
-    If rst.RecordCount > 1 Then Err.Raise 42, , "Multiple records in MSysResources table were found that matched name '" & m_Name & "' and extension '" & m_Extension & "' - Compact and repair database and try again."
+    If rst.RecordCount > 1 Then 
+        Err.Raise 42, , "Multiple records in MSysResources table were found that matched name '" & _
+                            m_Name & "' and extension '" & m_Extension & "' - Compact and repair database and try again."
+        CatchAny eelCritical, Err.Number & ":" & Err.Description, FunctionName & ":Too Many Themes:" & _
```
[the remark this answers is not in the document]
Comment (hecon5) [reply Version Control.accda.src/modules/clsDbTheme.bas:67]: good point; I struggled with this as to be explicit or to make it more human readable. Since it's a log file, I think you're right: human legibility is better here.

File: Version Control.accda.src/modules/clsDbTheme.bas [hunk 10]
```diff
@@ -11,7 +11,7 @@ Attribute VB_Exposed = False
 '---------------------------------------------------------------------------------------
 Option Compare Database
 Option Explicit
-
+Private Const moduleName As String = "clsDbTheme:"
```
[the remark this answers is not in the document]
Comment (hecon5) [reply Version Control.accda.src/modules/clsDbTheme.bas:14]: I looked high and low to avoid making a constant in a module that is the literal module name and should if possible be grabbed by, you know, the module name. Everything came up empty because it either isn't valid during execution, is not valid if you compile the module and run via accde, or some other ridiculous "of course Access / VBA does it that way and has to be different here". 

So, this won. I'll fix the case, though :)

[the remark this answers is not in the document]
Comment (joyfullservice) [reply Version Control.accda.src/modules/clsDbTheme.bas:14]: Yes, let's hope no one actually uses `ModuleName` as the literal name of their module!! That is definitely an exception that I don't think we are obligated to support.  😄

File: Version Control.accda.src/modules/clsDbTheme.bas (hunk 7 above)
[the remark this answers is not in the document]
Comment (joyfullservice) [reply Version Control.accda.src/modules/clsDbTheme.bas:65]: I checked the code history, it wasn't either of us. 😄 Let's go with the simpler direct logging of the error.

File: Version Control.accda.src/modules/modImportExport.bas (hunk 3 above)
Comment (joyfullservice) [reply Version Control.accda.src/modules/modImportExport.bas:158]: One nice thing here is that as we implement more and more error handling, this will only improve over time. When the global error level exceeds a certain threshold, it will _not auto-close the export screen_, so you will have a chance to stop and review the log file after you get back from the coffee break. (We can also update the completion Done to orange or red in color as another visual indicator that the operation completed with warnings/errors.)

Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:158]: Fair point; with that in mind; are you still of the mind to excise the error log dates?

File: Version Control.accda.src/modules/modImportExport.bas (hunk 5 above)
Comment (joyfullservice) [reply Version Control.accda.src/modules/modImportExport.bas:359]: Since all the case-specific error handling should be taking place in the object classes themselves, I think this could safely run as a single handler. We could even change the text to "processing " instead of the more specific verbs if we feel like it makes it to cumbersome to read. (The log file already tells you whether it was doing an import or a merge.)

File: Version Control.accda.src/modules/modImportExport.bas (hunk 3 above)
Comment (joyfullservice) [reply Version Control.accda.src/modules/modImportExport.bas:158]: I would still lean towards the single log file, and let the user save their own additional copies as needed. (Not create the date-based log files.)

File: Version Control.accda.src/modules/modImportExport.bas (hunk 5 above)
Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:359]: I agree; with that in mind, I propose: 

```VBA 
CatchAny eelError, Err.Number & ":" & Err.Description, ModuleName & ".build:FullBuild:" & blnFullBuild ",FastSave:" & _
    Options.UseFastSave & ",Category:" & LCase(cCategory.Category) & ",FileName:" & FSO.GetFileName(varFile), True, True

```

File: Version Control.accda.src/modules/clsDbTheme.bas [hunk 11]
```diff
@@ -102,6 +112,9 @@ End Sub
 '---------------------------------------------------------------------------------------
 '
 Private Sub IDbComponent_Import(strFile As String)
+    On Error Goto Err
+    Dim FunctionName As String
+    FunctionName = moduleName & "IDbComponent_Import:"
```
[the remark this answers is not in the document]
Comment (hecon5) [reply Version Control.accda.src/modules/clsDbTheme.bas:117]: should be fixed in a moment.

File: Version Control.accda.src/modules/clsDbTheme.bas [hunk 12]
```diff
@@ -183,6 +196,14 @@ Private Sub IDbComponent_Import(strFile As String)
...
     Set rstResources = Nothing
 
+Exit Sub
+
+Err:
+
+    CatchAny eelError, Err.Number & ":" & Err.Description, FunctionName & ":strFile:"  & strFile & ", blnIsFolder:" & _
+            blnIsFolder & ", strThemeFile:" & strThemeFile & ", Extension:" & m_Extension, True, True
+    Resume Next
+
```
[the remark this answers is not in the document]
Comment (hecon5) [reply Version Control.accda.src/modules/clsDbTheme.bas:206]: revised; let me know if this is what you were thinking.

File: Version Control.accda.src/modules/clsDbTheme.bas (hunk 11 above)
[the remark this answers is not in the document]
Comment (hecon5) [reply Version Control.accda.src/modules/clsDbTheme.bas:117]: removed

File: Version Control.accda.src/modules/modImportExport.bas (hunk 5 above)
Comment (joyfullservice) [reply Version Control.accda.src/modules/modImportExport.bas:359]: I am thinking something more like the following:

`CatchAny eelError, Err.Number & ": " & Err.Description & " while processing " & varFile, ModuleName & ".Build"`

Here's why. VBA doesn't really have a good mechanism to capture the global state of all variables in play at the time an error is encountered. (Think Locals window) And even if it did, I don't think we would want all of that dumped in an error log for a given error. Going with the principle of [keeping things simple](https://simpleprogrammer.com/kiss-one-best-practice-to-rule-them-all/), we really only want to log the details that are particularly meaningful and important. These are just clues to give us a head start when debugging a problem with the add-in. Most problems can be reproduced without much difficulty, so I don't really want to clutter up the code by passing in a lot of extra information that really isn't necessary to reproduce and resolve a problem.

In this case, we are looking at the following additional details beyond the error itself:

- FullBuild - We already have this in the log file from the context of the operation.
- UseFastSave - Already logged at the beginning of the operation.
- Category - This can be pretty easily identified from the file path.
- FileName - This also is obviously a part of the file path.

The file path is definitely an important detail. By logging the full path, we capture the category and file name. The other two options don't change during the operation and are already included in the log file.

Now one consideration here is that for the end user that is using this add-in for the first time and encounters an error exporting his or her database, having a more human-readable error may be helpful. On the flip side, our target audience is actual developers, not just casual users of Microsoft Access, so for them an actual file path might be more clear than the category name which is specific to this add-in.

Obviously both ways work fine, but as we go forward, I am just wanting to try to express some of the reasons behind the approach I am taking in various parts of the project.  😄  Simple. Clear. Readable. Consistent. Those are some of the qualities I am striving for. This project is (and will always be) a work in progress, but I personally feel that the better we lay the foundation, the better it will be for years to come.

So, to summarize, the way I would like to use the error logging is to pass the important and relevant details to the description parameter, and keep the source as "modModule.Function". I believe this will give us the important bits we need to debug problems while at the same time keeping the code fairly clean and light with the inline error handling.

File: Version Control.accda.src/modules/clsDbTheme.bas [hunk 13]
```diff
@@ -77,18 +73,21 @@ Private Sub IDbComponent_Export()
         Set rstAtc = rst!Data.Value
         If FSO.FileExists(strFile) Then DeleteFile strFile, True
         VerifyPath strFile
-        Perf.OperationStart "Extract Theme"
+        Perf.OperationStart "Export Theme"
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/clsDbTheme.bas:76]: Nice! Good idea to split the performance logging for the differences in Extracting vs. Exporting. 👍

File: Version Control.accda.src/modules/modImportExport.bas (hunk 5 above)
Comment (joyfullservice) [reply Version Control.accda.src/modules/modImportExport.bas:359]: Hang on... I just remembered something... We are _already_ logging the error number and description! (See clsLog.Error) The only thing we need to pass to the description is the _additional_ information we want to log.

`CatchAny eelError, "Error processing " & varFile, ModuleName & ".Build"`

That makes it even simpler!  😄

Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:359]: I like it! 

`strType = IIf(blnFullBuild, "Build", "Merge")` 

To be consistent, should this say `"Import", "Merge"` instead? Import/Build seem to be synonymous in this routine, and would ensure a consistent term use throughout.

Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:359]: To use this strType: `CatchAny eelError, "Error processing: " & varFile, ModuleName & ".Build:" & strType, True, True` 

What do you think?

File: Version Control.accda.src/modules/modImportExport.bas (hunk 6 above)
Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:399]: removed

File: Version Control.accda.src/modules/modImportExport.bas (hunk 5 above)
Comment (joyfullservice) [reply Version Control.accda.src/modules/modImportExport.bas:348]: > 
> `strType = IIf(blnFullBuild, "Build", "Merge")`
> 
> To be consistent, should this say `"Import", "Merge"` instead? Import/Build seem to be synonymous in this routine, and would ensure a consistent term use throughout.

Good point. Stepping back and thinking about this project in the big picture, it originally started out with the concept of importing source files. At that time, "import" was a proper and clear description. Fast forward to today, and the add-in has become more of a build tool that builds the entire database from source files. Similar to most other programming environments we have come to the (wonderful) place where we can now entirely _build_ a database from scratch. While it technically imports files to do this, the overall concept seems to be a "build" instead of an "import". This also seems to be consistent with other tools that "Build" a binary result from source files.

image

Looking forward, the "merge" idea is that existing source files will be _merged_ into the existing database project, which is different from building from scratch, and yet more involved then just importing files. (Some objects like database properties and references will be merged together with existing values, rather than strictly replacing them.)

So if we think of the entire process as a _build_, then the question is how to we describe the individual operations that take place during the overall build? Technically we are creating a blank database and importing objects from files, so _import_ may still be the best term for the individual items.

I definitely agree that it would be helpful to standardize on the terminology used. What do you think about the idea of making the distinction between build and import as it relates to whether we are referring to the operation as a whole, or to the processing of individual components?

File: Version Control.accda.src/modules/modImportExport.bas [hunk 14]
```diff
@@ -369,35 +358,32 @@ Public Sub Build(strSourceFolder As String, blnFullBuild As Boolean)
...
     Log.Add vbCrLf & Perf.GetReports, False
-    
-    If Log.ErrorLevel = eelNoError Then 
-        Log.SaveFile FSO.BuildPath(Options.GetExportFolder, IIf(blnFullBuild, "Import", "Merge") & ".log")
-    Else
-    ' Write log file to disk
-        Log.SaveFile FSO.BuildPath(Options.GetExportFolder, IIf(blnFullBuild, "Import-Error", "Merge-Error") & Format(Now, "-YYYY-MM-DD-hh.mm.ss") & ".log")
-    End If
 
+    Log.SaveFile FSO.BuildPath(Options.GetExportFolder, strType & ".log")
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:385]: Nice! 👍

File: Version Control.accda.src/modules/modImportExport.bas [hunk 15]
```diff
@@ -369,35 +358,32 @@ Public Sub Build(strSourceFolder As String, blnFullBuild As Boolean)
...
             Log.Add "Running " & Options.RunAfterBuild & "..."
-            Perf.OperationStart "RunAfterBuild"
+            Perf.OperationStart "RunAfter" & strType
             RunSubInCurrentProject Options.RunAfterBuild
             Perf.OperationEnd
         'End If
     End If
-
+    
+    CatchAny eelError, strType & " error in: " & varFile, ModuleName & ".Build:RunAfter" & strType, True, True
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:375]: `varFile` wouldn't have any meaning at this point... It would be fine to have a general error `CatchAny` at the end, but this should be updated per other discussions....

File: Version Control.accda.src/modules/modImportExport.bas (hunk 5 above)
Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:348]: I agree with the distinction. Especially when we're going to hand the routine name ".Build" to the logger, this might help steer someone to the cause.

File: Version Control.accda.src/modules/modImportExport.bas [hunk 16]
```diff
@@ -369,35 +358,32 @@ Public Sub Build(strSourceFolder As String, blnFullBuild As Boolean)
     If blnFullBuild Then
         If Options.RunAfterBuild <> vbNullString Then
             Log.Add "Running " & Options.RunAfterBuild & "..."
-            Perf.OperationStart "RunAfterBuild"
+            Perf.OperationStart "RunAfter" & strType
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:361]: Unless we are combining these with a CallByName on the build type, I don't know that we gain any benefit by not using the full name here. Also note that the RunAfterBuild option is different from the RunAfterMerge, since the two may require different internal processes afterwards. (I also observe that I had not fully finished this section of code.  😄)

File: Version Control.accda.src/modules/modImportExport.bas (hunk 5 above)
Comment (joyfullservice) [reply Version Control.accda.src/modules/modImportExport.bas:348]: > 
> 
> To use this strType: `CatchAny eelError, strType & "error in: " & varFile, ModuleName & ".Build", True, True`
> 
> What do you think?

Nice! Just be sure to add the space before " error in: ".  😄

File: Version Control.accda.src/modules/modImportExport.bas [hunk 17]
```diff
@@ -140,6 +138,7 @@ Public Sub ExportSource(blnFullExport As Boolean)
         Perf.OperationStart "RunAfterExport"
         RunSubInCurrentProject Options.RunAfterExport
         Perf.OperationEnd
+        CatchAny eelError, "Error exporting " & varFile, ModuleName & ".ExportSource:AfterExport", True, True
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:141]: See comment on similar line on Build function. (varFile does not have any meaning at this point in the process.)

File: Version Control.accda.src/modules/modImportExport.bas [hunk 18]
```diff
@@ -200,7 +192,7 @@ Public Sub Build(strSourceFolder As String, blnFullBuild As Boolean)
     Dim strText As String   ' Remove later
     
     ' The type of build will be used in various messages and log entries.
-    strType = IIf(blnFullBuild, "Build", "Merge")
+    strType = IIf(blnFullBuild, "Import", "Merge")
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:195]: I think we do want to keep this as "Build", per other discussions...

File: Version Control.accda.src/modules/modImportExport.bas [hunk 19]
```diff
@@ -338,7 +330,7 @@ Public Sub Build(strSourceFolder As String, blnFullBuild As Boolean)
         Else
             ' Show category header
             Log.Spacer Options.ShowDebug
-            Log.PadRight IIf(blnFullBuild, "Importing ", "Merging ") & LCase(cCategory.Category) & "...", , Options.ShowDebug
+            Log.PadRight strType & "ing " & LCase(cCategory.Category) & "...", , Options.ShowDebug
```
Comment (joyfullservice) [inline Version Control.accda.src/modules/modImportExport.bas:333]: Neat idea, but it could produce "Mergeing". Might be better to leave it as above for readability.

**Review by joyfullservice — COMMENTED** [review]
More notes added...

File: Version Control.accda.src/modules/modImportExport.bas (hunk 19 above)
Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:333]: Ah yeah, I had this from when leaning to import

Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:333]: Ah yeah, I had this from when leaning to import

File: Version Control.accda.src/modules/modImportExport.bas (hunk 18 above)
Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas]: done

File: Version Control.accda.src/modules/modImportExport.bas (hunk 16 above)
Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas]: Question on this: should this even be different? I'm trying to think of what routine you'd run after a build that you wouldn't run after a merge, since the point of the routine is to tidy up or set it to be ready/etc.

File: Version Control.accda.src/modules/modImportExport.bas (hunk 5 above)
Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:348]: fixed it even before you made the note, but wasn't as quick on the draw as you were :)

File: Version Control.accda.src/modules/modImportExport.bas (hunk 14 above)
Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas:385]: thanks!

File: Version Control.accda.src/modules/clsDbTheme.bas (hunk 13 above)
Comment (hecon5) [reply Version Control.accda.src/modules/clsDbTheme.bas:76]: thanks!

File: Version Control.accda.src/modules/modImportExport.bas (hunk 16 above)
Comment (joyfullservice) [reply Version Control.accda.src/modules/modImportExport.bas]: Yes, they could very likely be different. For example, after a full build you might run a routine to link in all the tables under the current user profile as described in a recent issue. But after a merge, you might want to keep the existing linked tables intact, and maybe instead just kick off a lighter unit testing procedure. 

Going forward the _Merge_ will be a key component of the sync, and I could certainly see scenarios where after making a tiny change to a module, I wouldn't want to run an entire startup routine, but rather just reopen my main form or something like that. I will admit that I don't personally have a [use case](https://martinfowler.com/bliki/Yagni.html) for it yet, but it just seems like the need is inevitable...

Comment (hecon5) [reply Version Control.accda.src/modules/modImportExport.bas]: fair point; half the features I've ever written were because a user offhand mentioned "oh, if only this did..." when I had spent hours pondering "did I miss anything" :)

**Comment by joyfullservice** [comment]
There are still a number of outstanding issues, but for the sake of expediency, let me go ahead and just merge this, and I will add another commit shortly with some additional cleanup work.

**Comment by hecon5** [comment]
Sorry for that; the theme fix got out of hand (my fault, entirely). Hopefully it's cruft that needs cleaned and not broken code :/

**Comment by joyfullservice** [comment]
All set! I didn't really take the time to test the theme import/export/extraction/rebuilding, as I think you were doing some of that in your project. Thanks again for your work!
